### PR TITLE
Introduce phased reconciliation process for CatalogSourceConfig objects

### PR DESCRIPTION
**Pull request by aravindhp**
- Fill in the CatalogSourceConfigStatus struct
- Introduce, 'Initial', 'Configuring', 'Succeeded' and 'Failed' phases
- Move the existing configuring logic in to the configuring reconciler

No new functionality has been added with these changes.

### CODE REVIEW

File: pkg/catalogsourceconfig/catalogsourceconfig.go [hunk 1]
```diff
@@ -2,8 +2,15 @@ package catalogsourceconfig
 
 import (
 	"github.com/operator-framework/operator-marketplace/pkg/datastore"
+	"github.com/operator-framework/operator-marketplace/pkg/phase"
 )
 
+// NewHandler returns an instance of the Handler interface
+// that can be used to reconcile an CatalogSourceConfig objects.
```
Comment (SamiSousa) [inline pkg/catalogsourceconfig/catalogsourceconfig.go]: Minor nit: I think it should be `a CatalogSourceConfig object` or `CatalogSourceConfig objects`

File: pkg/catalogsourceconfig/configuring.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+// TBD: Vendor in the folder once we require more than just these constants from
+// the OLM registry code
+const (
+	ConfigMapCRDName     = "customResourceDefinitions"
+	ConfigMapCSVName     = "clusterServiceVersions"
+	ConfigMapPackageName = "packages"
+)
+
+// NewConfiguringReconciler returns a Reconciler that reconciles a
+// CatalogSourceConfig object in the "Configuring" phase.
```
[the remark this answers is not in the document]
Comment (aravindhp) [reply pkg/catalogsourceconfig/configuring.go]: Not having "the" here is incorrect English in my opinion.

[the remark this answers is not in the document]
Comment (kevinrizza) [reply pkg/catalogsourceconfig/configuring.go]: I agree with Aravindh, I think it should be left as is.

File: pkg/catalogsourceconfig/configuring.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+		nextPhase = phase.GetNextWithMessage(phase.Failed, err.Error())
+		return
+	}
+
+	err = r.createCatalogSource(in, data)
+	if err != nil {
+		nextPhase = phase.GetNextWithMessage(phase.Failed, err.Error())
+		return
+	}
+
```
[the remark this answers is not in the document]
Comment (aravindhp) [reply pkg/catalogsourceconfig/configuring.go]: Good catch

File: pkg/catalogsourceconfig/configuring.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+	r.log.Info("The object has been successfully reconciled")
+	return
+}
+
+// createCatalogData constructs the ConfigMap data by reading the manifest
+// information of all packages from the datasource
```
[the remark this answers is not in the document]
Comment (aravindhp) [reply pkg/catalogsourceconfig/configuring.go]: I typically don't add a full stop for single line comments

File: pkg/catalogsourceconfig/configuring.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+		return err
+	}
+
+	catalogSource := newCatalogSource(cr, catalogConfigMap.Name)
+	err = sdk.Create(catalogSource)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		r.log.Errorf("Failed to create CatalogSource : %v", err)
+		return err
+	}
+	r.log.Infof("Created")
```
[the remark this answers is not in the document]
Comment (aravindhp) [reply pkg/catalogsourceconfig/configuring.go]: Sure. I will add a message for the ConfigMap also.

**Comment by aravindhp** [comment]
@tkashem @SamiSousa I have addressed your comment. Please review.

File: pkg/catalogsourceconfig/configuring.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (aravindhp) [reply pkg/catalogsourceconfig/configuring.go]: Looks like this is against golang [standards](https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences). I will fix it.

File: pkg/catalogsourceconfig/succeeded.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+// NewSucceededReconciler returns a Reconciler that reconciles a
+// CatalogSourceConfig object in the "Succeeded" phase.
+func NewSucceededReconciler(log *logrus.Entry) Reconciler {
+	return &succeededReconciler{
+		log: log,
+	}
+}
+
+// succeededReconciler is an implementation of Reconciler interface that
+// reconciles an CatalogSourceConfig object in "Succeeded" phase.
```
Comment (SamiSousa) [inline pkg/catalogsourceconfig/succeeded.go]: `in the "Succeeded" phase`

File: pkg/catalogsourceconfig/succeeded.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+type succeededReconciler struct {
+	log *logrus.Entry
+}
+
+// Reconcile reconciles an CatalogSourceConfig object that is in "Succeeded"
```
Comment (SamiSousa) [inline pkg/catalogsourceconfig/succeeded.go]: `in the "Succeeded" phase`

File: pkg/catalogsourceconfig/reconciler.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+// Reconciler is the interface that wraps the Reconcile method.
+//
+// Reconcile reconciles a particular phase of the CatalogSourceConfig object
+// and returns the next desired phase.
+//
+// in represents the original CatalogSourceConfig object received from the sdk
+// and before reconciliation has started.
+//
+// out represents the CatalogSourceConfig object after reconciliation has
+// completed and could be different from the original. The CatalogSourceConfig
```
Comment (SamiSousa) [inline pkg/catalogsourceconfig/reconciler.go]: I think that `and can be different` is better, but I'm not sure.

File: pkg/catalogsourceconfig/configuring.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (aravindhp) [reply pkg/catalogsourceconfig/configuring.go]: Done.

File: pkg/catalogsourceconfig/reconciler.go (hunk 8 above)
Comment (aravindhp) [reply pkg/catalogsourceconfig/reconciler.go]: `can` implies it is definitely different which is not the case here.

File: pkg/catalogsourceconfig/configuring.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tkashem) [reply pkg/catalogsourceconfig/configuring.go]: Sounds good, I saw both being used (line 38)

**Comment by aravindhp** [comment]
/hold
Until unit test case gets fixed

**Pull request merged**